### PR TITLE
Respect DefaultUserIsRestricted system default when creating new user

### PR DESCRIPTION
**Pull request by jpraet**
The DefaultUserIsRestricted system configuration options is not always enforced when creating a new user.

This PR configures the system defaults in user_model.CreateUser, and also enhances the CreateUserOverwriteOptions  to allow overwriting the system defaults when needed.

Note that the RegisterEmailConfirm and RegisterManualConfirm are also not enforced in most places.
Normalization of these settings has been taken out of the scope of this PR.

### CODE REVIEW

**Comment by jpraet** [comment]
There are a lot of places where `IsActive` is currently set to `true` instead of
`!(setting.Service.RegisterEmailConfirm || setting.Service.RegisterManualConfirm)`.
I'm not sure which of these need to remain `true` by design?

File: cmd/admin.go [hunk 1]
```diff
@@ -559,17 +564,26 @@ func runCreateUser(c *cli.Context) error {
...
 		Passwd:             password,
-		IsActive:           true,
 		IsAdmin:            c.Bool("admin"),
 		MustChangePassword: changePassword,
-		Theme:              setting.UI.DefaultTheme,
 	}
 
-	if err := user_model.CreateUser(u); err != nil {
+	overwriteDefault := &user_model.CreateUserOverwriteOptions{
+		IsActive:     util.OptionalBoolTrue,
```
Comment (jpraet) [inline cmd/admin.go:582]: OK to bypass setting.Service.RegisterEmailConfirm and setting.Service.RegisterManualConfirm here?

File: routers/api/v1/admin/user.go [hunk 2]
```diff
@@ -108,11 +108,17 @@ func CreateUser(ctx *context.APIContext) {
 		return
 	}
 
-	var overwriteDefault *user_model.CreateUserOverwriteOptions
+	overwriteDefault := &user_model.CreateUserOverwriteOptions{
+		IsActive: util.OptionalBoolTrue,
```
Comment (jpraet) [inline routers/api/v1/admin/user.go:112]: OK to bypass setting.Service.RegisterEmailConfirm and setting.Service.RegisterManualConfirm here?

File: routers/web/admin/users.go [hunk 3]
```diff
@@ -125,10 +125,14 @@ func NewUserPost(ctx *context.Context) {
 		Name:      form.UserName,
 		Email:     form.Email,
 		Passwd:    form.Password,
-		IsActive:  true,
 		LoginType: auth.Plain,
 	}
 
+	overwriteDefault := &user_model.CreateUserOverwriteOptions{
+		IsActive:   util.OptionalBoolTrue,
```
Comment (jpraet) [inline routers/web/admin/users.go:132]: OK to bypass setting.Service.RegisterEmailConfirm and setting.Service.RegisterManualConfirm here?

File: routers/web/auth/oauth.go [hunk 4]
```diff
@@ -867,19 +868,21 @@ func SignInOAuthCallback(ctx *context.Context) {
...
+				Name:        getUserName(&gothUser),
+				FullName:    gothUser.Name,
+				Email:       gothUser.Email,
+				LoginType:   auth.OAuth2,
+				LoginSource: authSource.ID,
+				LoginName:   gothUser.UserID,
+			}
+
+			overwriteDefault := &user_model.CreateUserOverwriteOptions{
+				IsActive: util.OptionalBoolOf(!setting.OAuth2Client.RegisterEmailConfirm),
```
Comment (jpraet) [inline routers/web/auth/oauth.go:880]: OK to bypass setting.Service.RegisterManualConfirm here?

File: services/auth/reverseproxy.go [hunk 5]
```diff
@@ -105,11 +106,15 @@ func (r *ReverseProxy) newUser(req *http.Request) *user_model.User {
...
-		Name:     username,
-		Email:    email,
-		IsActive: true,
+		Name:  username,
+		Email: email,
 	}
-	if err := user_model.CreateUser(user); err != nil {
+
+	overwriteDefault := user_model.CreateUserOverwriteOptions{
+		IsActive: util.OptionalBoolTrue,
```
Comment (jpraet) [inline services/auth/reverseproxy.go:114]: OK to bypass setting.Service.RegisterEmailConfirm and setting.Service.RegisterManualConfirm here?

File: services/auth/source/ldap/source_authenticate.go [hunk 6]
```diff
@@ -85,19 +86,21 @@ func (source *Source) Authenticate(user *user_model.User, userName, password str
...
+		FullName:    composeFullName(sr.Name, sr.Surname, sr.Username),
+		Email:       sr.Mail,
+		LoginType:   source.authSource.Type,
+		LoginSource: source.authSource.ID,
+		LoginName:   userName,
+		IsAdmin:     sr.IsAdmin,
+	}
+	overwriteDefault := &user_model.CreateUserOverwriteOptions{
+		IsRestricted: util.OptionalBoolOf(sr.IsRestricted),
+		IsActive:     util.OptionalBoolTrue,
```
Comment (jpraet) [inline services/auth/source/ldap/source_authenticate.go:100]: OK to bypass setting.Service.RegisterEmailConfirm and setting.Service.RegisterManualConfirm here?

File: services/auth/source/ldap/source_sync.go [hunk 7]
```diff
@@ -102,19 +103,21 @@ func (source *Source) Sync(ctx context.Context, updateExisting bool) error {
...
+				FullName:    fullName,
+				LoginType:   source.authSource.Type,
+				LoginSource: source.authSource.ID,
+				LoginName:   su.Username,
+				Email:       su.Mail,
+				IsAdmin:     su.IsAdmin,
+			}
+			overwriteDefault := &user_model.CreateUserOverwriteOptions{
+				IsRestricted: util.OptionalBoolOf(su.IsRestricted),
+				IsActive:     util.OptionalBoolTrue,
```
Comment (jpraet) [inline services/auth/source/ldap/source_sync.go:117]: OK to bypass setting.Service.RegisterEmailConfirm and setting.Service.RegisterManualConfirm here?

edit: my guess would be that this is OK here, as this sync is triggered when cron.sync_external_users is enabled, indicating the gitea admin intends the user accounts to be created automatically based on the LDAP data.

File: services/auth/source/pam/source_authenticate.go [hunk 8]
```diff
@@ -58,10 +59,12 @@ func (source *Source) Authenticate(user *user_model.User, userName, password str
 		LoginType:   auth.PAM,
 		LoginSource: source.authSource.ID,
 		LoginName:   userName, // This is what the user typed in
-		IsActive:    true,
+	}
+	overwriteDefault := &user_model.CreateUserOverwriteOptions{
+		IsActive: util.OptionalBoolTrue,
```
Comment (jpraet) [inline services/auth/source/pam/source_authenticate.go:64]: OK to bypass setting.Service.RegisterEmailConfirm and setting.Service.RegisterManualConfirm here?

File: services/auth/source/smtp/source_authenticate.go [hunk 9]
```diff
@@ -74,10 +74,12 @@ func (source *Source) Authenticate(user *user_model.User, userName, password str
 		LoginType:   auth_model.SMTP,
 		LoginSource: source.authSource.ID,
 		LoginName:   userName,
-		IsActive:    true,
+	}
+	overwriteDefault := &user_model.CreateUserOverwriteOptions{
+		IsActive: util.OptionalBoolTrue,
```
Comment (jpraet) [inline services/auth/source/smtp/source_authenticate.go:79]: OK to bypass setting.Service.RegisterEmailConfirm and setting.Service.RegisterManualConfirm here?

edit: my guess would be that it's ok to bypass RegisterEmailConfirm here, as the successful authentication against the trusted SMTP server implies a valid email adress? RegisterManualConfirm should probably be taken into account though.

File: cmd/admin.go (hunk 1 above)
Comment (lunny) [reply cmd/admin.go:582]: I'm OK for that. Since this is admin to create a user.

File: routers/api/v1/admin/user.go (hunk 2 above)
Comment (jpraet) [reply routers/api/v1/admin/user.go:112]: OK as per https://github.com/go-gitea/gitea/pull/19310#discussion_r842302420

File: routers/web/admin/users.go (hunk 3 above)
Comment (jpraet) [reply routers/web/admin/users.go:132]: OK as per https://github.com/go-gitea/gitea/pull/19310#discussion_r842302420

**Comment by jpraet** [comment]
Could anyone advise on where setting.Service.RegisterEmailConfirm and/or setting.Service.RegisterManualConfirm can be bypassed and where they cannot?

**Comment by lunny** [comment]
> Could anyone advise on where setting.Service.RegisterEmailConfirm and/or setting.Service.RegisterManualConfirm can be bypassed and where they cannot?

I think gitea command line could ignore these options. Otherwise they should be obey.

**Comment by lafriks** [comment]
I don't really think external authentication sources should obey to these rules also, probably they could have option to them to added then to activate user or not but currently I don't think that's a good idea to have to activate them

**Comment by jpraet** [comment]
My initial reason for this PR was that DefaultUserIsRestricted is currently not taken into account in multiple places where new user accounts are created.

When discussing this on discord @zeripath mentioned _"you might wanna check isManualConfirm too - appears I may have missed some"_, so I broadened the scope to also look into RegisterEmailConfirm and RegisterManualConfirm as well.

But it looks like there's no consensus as to where these settings should apply?

**Comment by jpraet** [comment]
Should I maybe leave the RegisterEmailConfirm and RegisterManualConfirm settings AS-IS, and only address DefaultUserIsRestricted in this PR?

**Comment by lunny** [comment]
> Should I maybe leave the RegisterEmailConfirm and RegisterManualConfirm settings AS-IS, and only address DefaultUserIsRestricted in this PR?

I'm OK for that.

File: cmd/admin.go [hunk 10]
```diff
@@ -559,17 +564,26 @@ func runCreateUser(c *cli.Context) error {
...
+	}
+
 	u := &user_model.User{
 		Name:               username,
 		Email:              c.String("email"),
 		Passwd:             password,
-		IsActive:           true,
 		IsAdmin:            c.Bool("admin"),
 		MustChangePassword: changePassword,
-		Theme:              setting.UI.DefaultTheme,
```
Comment (lafriks) [inline cmd/admin.go:569]: Why default theme is removed?

Comment (jpraet) [reply cmd/admin.go:569]: Already applied by default in CreateUser:
https://github.com/go-gitea/gitea/blob/5574ad119beebbafeaef3dd46c3f3de022e4ab1b/models/user/user.go#L646

File: services/auth/sspi_windows.go [hunk 11]
```diff
@@ -187,17 +188,20 @@ func (s *SSPI) shouldAuthenticate(req *http.Request) (shouldAuth bool) {
...
+		Passwd:          gouuid.New().String(),
+		Language:        cfg.DefaultLanguage,
+		UseCustomAvatar: true,
+		Avatar:          avatars.DefaultAvatarLink(),
+	}
+	emailNotificationPreference := user_model.EmailNotificationsDisabled
+	overwriteDefault := &user_model.CreateUserOverwriteOptions{
+		IsActive:                     util.OptionalBoolOf(cfg.AutoActivateUsers),
+		KeepEmailPrivate:             util.OptionalBoolTrue,
+		EmailNotificationsPreference: &emailNotificationPreference,
```
Comment (jpraet) [inline services/auth/sspi_windows.go:202]: This actually changes the current behavior though, as previously the EmailNotificationsPreference would get overwritten by the setting.Admin.DefaultEmailNotification in CreateUser. Is EmailNotificationsDisabled what was intended here for sppi_windows?

Comment (zeripath) [reply services/auth/sspi_windows.go:202]: I think the previous behaviour was likely a mistake.

**Review by zeripath — APPROVED** [review]

**Review by wxiaoguang — APPROVED** [review]

**Pull request merged**